modules/base/rendering/renderabletimevaryingsphere.cpp

```diff
@@ -0,0 +1,554 @@
+    computeSequenceEndTime();
+    loadTexture();
+}
+
+void RenderableTimeVaryingSphere::deinitializeGL() {
+    _texture = nullptr;
+
+    BaseModule::ProgramObjectManager.release(
+        ProgramName,
+        [](ghoul::opengl::ProgramObject* p) {
+            global::renderEngine->removeRenderProgram(p);
+        }
+    );
+    _files.clear();
+    _shader = nullptr;
+}
+
+void RenderableTimeVaryingSphere::render(const RenderData& data, RendererTasks&) {
+    Orientation orientation = static_cast<Orientation>(_orientation.value());
+
+    glm::dmat4 modelTransform =
+        glm::translate(glm::dmat4(1.0), data.modelTransform.translation) *
+        glm::dmat4(data.modelTransform.rotation) *
+        glm::scale(glm::dmat4(1.0), glm::dvec3(data.modelTransform.scale));
+
+    glm::dmat3 modelRotation = data.modelTransform.rotation;
+
+    _shader->activate();
+
+    glm::mat4 modelViewProjection = data.camera.projectionMatrix() *
+                             glm::mat4(data.camera.combinedViewMatrix() * modelTransform);
+    _shader->setUniform(_uniformCache.modelViewProjection, modelViewProjection);
+
+    glm::mat3 modelViewRotation = glm::mat3(
+        glm::dmat3(data.camera.viewRotationMatrix()) * modelRotation
+    );
+    _shader->setUniform(_uniformCache.modelViewRotation, modelViewRotation);
+
+    float adjustedOpacity = _opacity;
+
+    if (!_disableFadeInDistance) {
+        if (_fadeInThreshold > -1.0) {
+            const float logDistCamera = glm::log(static_cast<float>(
+                glm::distance(data.camera.positionVec3(), data.modelTransform.translation)
+            ));
+            const float startLogFadeDistance = glm::log(_size * _fadeInThreshold);
+            const float stopLogFadeDistance = startLogFadeDistance + 1.f;
+
+            if (logDistCamera > startLogFadeDistance &&
+                logDistCamera < stopLogFadeDistance)
+            {
+                const float fadeFactor = glm::clamp(
+                    (logDistCamera - startLogFadeDistance) /
+                        (stopLogFadeDistance - startLogFadeDistance),
+                    0.f,
+                    1.f
+                );
+                adjustedOpacity *= fadeFactor;
+            }
+            else if (logDistCamera <= startLogFadeDistance) {
+                adjustedOpacity = 0.f;
+            }
+        }
+
+        if (_fadeOutThreshold > -1.0) {
+            const float logDistCamera = glm::log(static_cast<float>(
+                glm::distance(data.camera.positionVec3(), data.modelTransform.translation)
+            ));
+            const float startLogFadeDistance = glm::log(_size * _fadeOutThreshold);
+            const float stopLogFadeDistance = startLogFadeDistance + 1.f;
+
+            if (logDistCamera > startLogFadeDistance && 
+                logDistCamera < stopLogFadeDistance)
+            {
+                const float fadeFactor = glm::clamp(
+                    (logDistCamera - startLogFadeDistance) /
```

Comment:
To make this code a little easier to read, maybe create a variable for the `(logDistCamera - startLogFadeDistance) / (stopLogFadeDistance - startLogFadeDistance)`. Then the fadeFactor claming could probably be done on one line